R/read_vpts.R

```diff
@@ -1,241 +1,124 @@
-    data$datetime <- as.POSIXct(paste(data$Date, sprintf("%04d", data$Time),
-                                      sep = ""
-    ),
-    format = "%Y%m%d%H%M",
-    tz = "UTC"
-    )
-  }
-
-  # add profile_index to identify consecutive profiles
-  data$new_profile_starts <- c(T, (data$height[-1] - data$height[-length(data$height)]) < 0)
-  data$profile_index <- NA
-  profile_index <- NULL # define profile_index to suppress devtools::check warning in next line
-  data[which(data$new_profile_starts), "profile_index"] <- 1:length(which(data$new_profile_starts))
-  data <- tidyr::fill(data, profile_index)
+  # Read files
+  data <- switch(extension,
+    csv = read_vpts_csv(files),
+    gz = read_vpts_csv(files),
+    h5 = read_vpts_hdf5(files)
+  )
 
-  data$new_profile_starts <- NULL
-  data$Date <- NULL
-  data$Time <- NULL
+  data
+}
 
-  # sort
-  data <- data[with(data, order(datetime, profile_index, height)), ]
+#' Read time series of vertical profiles (`vpts`) from VPTS CSV file(s)
+#'
+#' @inheritParams read_vpts
+#' @param df If `TRUE` returns data as data frame rather than `vpts`
+#'   object.
+#' @return `vpts` object.
+#' @noRd
+read_vpts_csv <- function(files, df = FALSE) {
+  # Create Frictionless Data Package
+  package <- frictionless::create_package()
+  schema <- "https://raw.githubusercontent.com/enram/vpts-csv/main/vpts-csv-table-schema.json"
+  package <- frictionless::add_resource(
+    package,
+    "vpts",
+    data = files,
+    schema = schema
+  )
 
-  # split into profiles
-  data <- split(data, data$profile_index)
-  names(data) <- NULL
+  # Read resource (compares data with schema and binds rows of all files)
+  data <- frictionless::read_resource(package, "vpts")
 
-  # verify that profiles can be flattened
-  datadim <- sapply(1:length(data), function(x) dim(data[[x]]))
+  # Convert data
+  data <- dplyr::mutate(
+    data,
+    radar = as.factor(radar),
+    source_file = as.factor(source_file)
+  )
 
-  if (length(unique(datadim[1, ])) > 1) {
-    mostFrequent <- sort(table(datadim[1, ]), decreasing = TRUE)[1]
-    if (mostFrequent <= 1) {
-      stop("Profiles are of unequal altitudinal dimensions, unable to merge")
-    }
-    mostFrequentNBins <- as.integer(names(mostFrequent))
-    warning(paste(
-      "Profiles are of unequal altitudinal dimensions or",
-      "contain duplicates. Discarding", length(data) - mostFrequent,
-      "of", length(data), "profiles, restricting to",
-      mostFrequentNBins, "altitude bins."
-    ))
-    data <- data[datadim[1, ] == mostFrequentNBins]
+  # Return data as data frame
+  if (df) {
+    return(data)
   }
 
-  # strip the datetime field
-  datetime <- .POSIXct(sapply(
-    1:length(data),
-    function(x) {
-      data[[x]]$datetime[1]
-    }
-  ),
-  tz = "UTC"
-  )
-  data <- lapply(
-    data,
-    function(x) {
-      x["datetime"] <- NULL
-      x["profile_index"] <- NULL
-      x
-    }
+  # The following steps convert the data to a vpts object
```

Comment:
Would it be an idea to put this data frame to vpts conversion into a separate function, that would enable reuse by people that get the data for example out of a database. I'm not sure as this also touches on the new vpts format. This might also relate to #555

Reply:
Yeah, but it becomes moot the moment we move vpts objects to dataframes _in_ bioRad (#568).